src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    [ExportWorkspaceService(typeof(IUnusedReferencesService), ServiceLayer.Default), Shared]
+    internal class UnusedReferencesService : IUnusedReferencesService
+    {
+        // We set this processing order because we want to favor transitive references when possible.
+        // For instance we process Projects before Packages, since a particular Package could be brought
+        // in transitively by a Project reference.
+        private static readonly ReferenceType[] _processingOrder = new[]
+        {
+            ReferenceType.Project,
+            ReferenceType.Package,
+            ReferenceType.Assembly
+        };
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public UnusedReferencesService()
+        {
+        }
+
+        public async Task<ImmutableArray<ReferenceInfo>> GetUnusedReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceInfo> references,
+            CancellationToken cancellationToken)
+        {
+            // Create a lookup of used assembly paths
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            if (compilation is null)
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            HashSet<string> usedAssemblyLookup = new(compilation.GetUsedAssemblyReferences(cancellationToken)
+                .Select(reference => reference.Display)
+                .OfType<string>());
```

Comment:
This is grabbing file paths? You should be able to do OfType<PortableExecutableReference>() and from there grab the file path. That's probably what the Display is going to be anyways, but at least it's more explicit.